src/cargo/core/resolver/encode.rs

```diff
@@ -596,28 +597,39 @@ impl<'de> de::Deserialize<'de> for EncodableSourceId {
         D: de::Deserializer<'de>,
     {
         let s = String::deserialize(d)?;
-        let sid = SourceId::from_url(&s).map_err(de::Error::custom)?;
-        Ok(EncodableSourceId {
-            inner: sid,
-            encoded: false,
-        })
+        Ok(EncodableSourceId::new(s).map_err(de::Error::custom)?)
     }
 }
 
 impl std::hash::Hash for EncodableSourceId {
     fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
-        self.inner.hash(state)
+        self.kind.hash(state);
+        self.url.hash(state);
     }
 }
 
 impl std::cmp::PartialEq for EncodableSourceId {
     fn eq(&self, other: &Self) -> bool {
-        self.inner == other.inner
+        self.url == other.url && self.kind == other.kind
```

Comment:
Should we do `kind` and then `url`?  That is a faster check and also the order we do it elsewhere